File: go/0482_license_key_formatting.go
import "bytes"

func licenseKeyFormatting(S string, K int) string {
	var b bytes.Buffer
	cnt := 0
	for i := len(S) - 1; i >= 0; i-- {
		ch := S[i]
		if ch == '-' {
			continue
		} else if ch >= 'a' && ch <= 'z' {
			ch = byte(int(ch) + (int('Z') - int('z')))
		}
		b.WriteByte(ch)
		cnt++
		if cnt == K {
			cnt = 0
			b.WriteByte('-')
		}
	}

	res := b.Bytes()
	// we may have added unnecessary '-' to the last, so omit it if any
	if len(res) > 0 && res[len(res)-1] == '-' {
		res = res[:len(res)-1]
	}
	return reverse(res)
}

func reverse(b []byte) string {
	for l, r := 0, len(b)-1; l < r; l, r = l+1, r-1 {
		b[l], b[r] = b[r], b[l]
	}
	return string(b)
}

File: python/0482_license_key_formatting.py
class Solution:
    def licenseKeyFormatting(self, S: str, K: int) -> str:
        count = 0
        res = []
        
        i = len(S) - 1
        while i >= 0:
            if count == K:
                if S[i] == '-':
                    res.append(S[i])
                else:
                    res.append('-')
                count = 0
                continue
            else:
                if S[i] == '-':
                    i -= 1
                else:
                    res.append(S[i].upper())
                    i -= 1
                    count += 1
        if res and res[-1] == '-':
            res = res[:-1]
        res.reverse()
        return ''.join(res)